packages/firestore/test/unit/local/index_free_query_engine.test.ts

```diff
@@ -0,0 +1,362 @@
+      }
+    );
+  }
+
+  /** Adds the provided documents to the remote document cache.  */
+  function addDocument(...docs: MaybeDocument[]): Promise<void> {
+    return persistence.runTransaction('addDocument', 'readwrite', txn => {
+      const changeBuffer = remoteDocumentCache.newChangeBuffer();
+      for (const doc of docs) {
+        changeBuffer.addEntry(doc, doc.version);
+      }
+      return changeBuffer.apply(txn);
+    });
+  }
+
+  async function expectIndexFreeQuery<T>(op: () => Promise<T>): Promise<T> {
+    try {
+      localDocuments.expectIndexFreeExecution = true;
+      return await op();
+    } finally {
+      localDocuments.expectIndexFreeExecution = undefined;
+    }
+  }
+
+  async function expectFullCollectionQuery<T>(
+    op: () => Promise<T>
+  ): Promise<T> {
+    try {
+      localDocuments.expectIndexFreeExecution = false;
+      return await op();
+    } finally {
+      localDocuments.expectIndexFreeExecution = undefined;
+    }
+  }
+
+  function runQuery(query: Query, queryData: QueryData): Promise<DocumentSet> {
+    assert(
+      localDocuments.expectIndexFreeExecution !== undefined,
+      'Encountered runQuery() call not wrapped in expectIndexFreeQuery()/expectFullCollectionQuery()'
+    );
+
+    return persistence.runTransaction('runQuery', 'readonly', txn => {
+      return queryCache
+        .getMatchingKeysForTargetId(txn, TEST_TARGET_ID)
+        .next(remoteKeys => {
+          return queryEngine
+            .getDocumentsMatchingQuery(txn, query, queryData, remoteKeys)
+            .next(docs => {
+              const view = new View(query, remoteKeys);
+              const viewDocChanges = view.computeDocChanges(docs);
+              return view.applyChanges(
+                viewDocChanges,
+                /*updateLimboDocuments=*/ true
+              ).snapshot!.docs;
+            });
+        });
+    });
+  }
+
+  beforeEach(async () => {
+    persistence = await testMemoryEagerPersistence();
+    queryCache = persistence.getQueryCache();
+    queryEngine = new IndexFreeQueryEngine();
+
+    remoteDocumentCache = persistence.getRemoteDocumentCache();
+
+    localDocuments = new TestLocalDocumentsView(
+      remoteDocumentCache,
+      persistence.getMutationQueue(User.UNAUTHENTICATED),
+      new MemoryIndexManager()
+    );
+    queryEngine.setLocalDocumentsView(localDocuments);
+  });
+
+  it('uses target mapping for initial view', async () => {
+    const query = Query.atPath(path('coll')).addFilter(
+      filter('matches', '==', true)
+    );
+    const queryData = testQueryData(query, /* hasLimboFreeSnapshot= */ true);
+
+    await addDocument(MATCHING_DOC_A, MATCHING_DOC_B);
+    await persistQueryMapping(MATCHING_DOC_A.key, MATCHING_DOC_B.key);
+
+    const docs = await expectIndexFreeQuery(() => runQuery(query, queryData));
+
+    verifyResult(docs, [MATCHING_DOC_A, MATCHING_DOC_B]);
+  });
+
+  it('filters non-matching initial results', async () => {
```

Comment:
I'm not sure what "initial results" refers to...  should this be "filters non-matching changes since initial results" by chance?